tfx/orchestration/experimental/kubernetes/kubernetes_dag_runner.py

```diff
@@ -0,0 +1,378 @@
+        component=component,
+        pipeline_info=pipeline.pipeline_info,
+        driver_args=driver_args,
+        metadata_connection=metadata_connection,
+        beam_pipeline_args=pipeline.beam_pipeline_args,
+        additional_pipeline_args=pipeline.additional_pipeline_args,
+        component_config=component_config)
+    self._component_id = component.id
+
+  def run_component(self) -> None:
+    absl.logging.info('Component %s is running.', self._component_id)
+    self._component_launcher.launch()
+    absl.logging.info('Component %s is finished.', self._component_id)
+
+
+class KubernetesDagRunner(tfx_runner.TfxRunner):
+  """Tfx runner on Kubernetes."""
+
+  def __init__(self,
+               config: Optional[pipeline_config.PipelineConfig] = None):
+    """Initializes BeamDagRunner as a TFX orchestrator.
+
+    Args:
+      config: Optional pipeline config for customizing the launching of each
+        component. Defaults to pipeline config that supports
+        InProcessComponentLauncher and KubernetesComponentLauncher.
+    """
+    if config is None:
+      config = pipeline_config.PipelineConfig(
+          supported_launcher_classes=[
+              in_process_component_launcher.InProcessComponentLauncher,
+              kubernetes_component_launcher.KubernetesComponentLauncher,
+          ],
+      )
+    super(KubernetesDagRunner, self).__init__(config)
+
+  def run(self, pipeline: tfx_pipeline.Pipeline) -> None:
+    """
+    Args:
+      pipeline: Logical pipeline containing pipeline args and components.
+    """
+    if not pipeline.pipeline_info.run_id:
+      pipeline.pipeline_info.run_id = datetime.datetime.now().isoformat()
+
+    if not is_inside_cluster():
+      self._run_as_kubernetes_job(pipeline)
+      return
+
+    # TODO(ericlege) support running components in parallel
+    ran_components = set()
+
+    # Runs component in topological order
+    for component in pipeline.components:
+
+      # verify that components are in topological order
+      if hasattr(component, 'upstream_nodes') and component.upstream_nodes:
+        for upstream_node in component.upstream_nodes:
+          assert upstream_node in ran_components, ('Components is not in '
+                                                   'topological order')
+
+      (component_launcher_class,
+       component_config) = config_utils.find_component_launch_info(
+           self._config, component)
+
+      # Check if the component is launchable as a containerComponent.
+      # If not, wrap the component to a containerComponent.
+      if kubernetes_component_launcher.KubernetesComponentLauncher.can_launch(
+          component.executor_spec, component_config):
+        LaunchAsContainerComponent(component,
+                                   component_launcher_class,
+                                   component_config,
+                                   pipeline).run_component()
+      else:
```

Comment:
We should do a check that this can be launched by the InProcessComponentLauncher (raising a ValueError with an appropriate message if not).

Reply:
Agreed.